src/consensus.rs

```diff
@@ -491,58 +429,53 @@ impl Consensus {
             .sum::<u64>()
     }
 
-    /// Verify that the certificate signers are legit for proposal
-    /// and update current proposal if the quorum is valid
-    /// precondition: verification of voting power is already made on quorum
-    fn verify_quorum_and_catchup(
-        &mut self,
-        consensus_proposal_hash: &ConsensusProposalHash,
-        prepare_quorum_certificate: &QuorumCertificate,
-    ) -> Result<(), Error> {
-        // Verify that the Confirm is for the correct proposal
-        let proposal: ConsensusProposal = match self
-            .verify_consensus_proposal_hash(consensus_proposal_hash)
-        {
-            true => self.bft_round_state.consensus_proposal.clone(),
-            false => {
-                // If the verification failed, we might have missed a slot
-                if let Some(proposal) = self
-                    .buffered_invalid_proposals
-                    .remove(consensus_proposal_hash)
-                {
-                    warn!("Confirm refers to proposal hash I didn't vote for because I might have missed a slot. Restoring this proposal as valid if quorum signers are part of the consensus.");
-                    proposal
-                } else {
-                    bail!("Confirm refers to an unknwon proposal hash.");
-                }
+    /// Verify that:
+    ///  - the quorum certificate is for the given message.
+    ///  - the signatures are above 2f+1 voting power.
+    ///
+    /// This ensures that we can trust the message.
+    fn verify_quorum_certificate<T: bincode::Encode>(
+        &self,
+        message: T,
+        quorum_certificate: &QuorumCertificate,
+    ) -> Result<()> {
+        // Construct the expected signed message
+        let expected_signed_message = SignedWithKey {
+            msg: message,
+            signature: quorum_certificate.signature.clone(),
+            validators: quorum_certificate.validators.clone(),
+        };
+
+        match BlstCrypto::verify(&expected_signed_message) {
+            Ok(res) if !res => {
+                //self.metrics.confirm_error("qc_invalid"); todo
+                bail!("Quorum Certificate received is invalid")
             }
+            Err(err) => bail!("Quorum Certificate verification failed: {}", err),
+            _ => {}
         };
 
-        if proposal.slot < self.bft_round_state.slot {
-            bail!(
-                "Quorum for a passed slot {} while we are in slot {}",
-                proposal.slot,
-                self.bft_round_state.slot
-            );
-        }
-        if proposal.slot == self.bft_round_state.slot && proposal.view < self.bft_round_state.view {
-            bail!(
-                "Quorum for a passed view {}:{} while we are in slot {}:{}",
```

Comment:
On n'a plus besoin de check ça ?

Reply:
Nope c'est inclus for free dans le check de la signature, puisque le message signé inclue la view et le slot.
Maintenant je suis d'accord que c'est vraiment pas très explicite, il faudrait le re-rendre apparent plus tard IMO.

Reply:
le message & la signature peuvent être valides s'ils sont cohérent entre eux, ça ne permet pas au node de savoir s'ils sont cohérents avec son état internet et que c'est bien le message attendu

Reply:
oui ok en effet, même histoire, en ajoutant slot/view aux messages on s'y retrouvera